Add hub.yaml proposal for themes (#422)
obsidian-community/obsidian-hub: 00 - Contribute to the Obsidian Hub/03 Contributor Notes/03.02 Design Decisions/hub.yaml proposal for themes.md

---
publish: true
---

> # `hub.yaml` proposal for theme designers

Reviewer: just a precaution, but sometimes you wrote `hub.yml`, sometimes `hub.yaml` – we should be consistent about that, or better, the scripter of thingies should look for both, just in case. Cause I am pretty sure some theme devs gonna write it one way, others the other way.

Author: Huh weird, search and replace didn't find any `yml` strings in the file

        "branch": "main"
    },
   ```
- Your `obsidian.css` file to determine if you use the Style settings plugin and to see if you've listed any [compatible plugins](https://publish.obsidian.md/hub/04+-+Guides%2C+Workflows%2C+%26+Courses/Guides/Adding+plugin+compatibility+for+themes+to+the+Obsidian+Hub).

This is covers the very basic information for most themes, but
- is not very user-friendly when trying to find themes that have X (which themes support asides? highlights?) and users have to try out the themes to find out
- limits you to a single screenshot
- requires manual edits on your side if you want to add text or other features not mentioned in style settings or list sponsorship information
	- content you add to the hub is  outside to your theme repo and will easily get out of date
	- adding or updating any content you do add to the hub requires you to go through the fork and PR workflow, which adds friction and makes it less likely that you do either
- requires manual work in the hub  repo when there are manual changes to make sure the scripts haven't removed manually added content and handling other edge cases (at least for now)

## Proposal

> Our proposal is to add an optional `hub.yaml` file at the root of theme repos to add some of this information. Note that all entries are optional, so maintenance and effort can be managed by each theme designer based on the their theme's complexity and requirements. We'll cover some of the basics with examples below, but you can find the full file with all the options further down in the [[#Template]] section.

Reviewer 2: ```suggestion
Our proposal is to add an optional `hub.yaml` file at the root of theme repos to add some of this information. Note that all entries are optional, so maintenance and effort can be managed by each theme designer based on their theme's complexity and requirements. We'll cover some of the basics with examples below, but you can find the full file with all the options further down in the [[#Template]] section.
```

### Link to the theme's documentation

The contents of the `hub.yaml` for a theme that **only** wants to add a link to the theme's documentation in its hub note would be as simple as:

```yaml
docs: https://minimal.guide/ # Link to the theme documentation
```

### Hub note contents

Similarly, if you wanted to have some control on the contents of your theme note, you can use the `hub_content` key to list which markdown documents to "import" to the hub. Generally, to avoid having to maintain two sets of docs (one for your theme and one for the hub), you could reuse a markdown file with an overview that has links to your docs.

For example, with the contents of `hub.yaml` as show below:
```yaml
hub_content: # List of markdown documents to be included as is to the hub relative to the root of the repo.
	- docs/theme_overview.md
```

And the contents of `theme_overview.md`:

> ```markdown

Reviewer 2: Any thoughts about relative links in the markdown? I suspect we will soon end up having to parse the imported markdown files and add URLs where the markdown had internal links within the theme's repo.

Author: I'd start with "as is", no parsing or adding any other links. If there are links, we should recommend that they are always external (better for reuse) or use `[[wikilinks]]` of notes in the hub (if they are making such file specifically for us).

Reviewer 2: Sounds good. If that's not already in the doc, might it be possible to include that?

## My theme overview

This is a small description of my theme. You can find more information about the following features in the documentation:
- [Callouts](www.mytheme.com/callouts)
- [Checklists](www.mytheme.com/checklists)

Report issues by clicking [here](www.github.com/octocat/mytheme/issues).
```

### Theme features

> The main goal of adding a list of features is to help compare themes based on what they offer, before one has to try them out. Keep in mind that themes (especially large ones) have many many features, and it's not our intention to make an exhaustive list of **all** the things your theme can do. Instead, ==ideally the features you list are a way to pitch to (new) users what makes your theme unique==. For that, the idea is to parse the `feature` key to identify some of the frequently sought-after features with a controlled vocabulary and automatically add links to pre-defined categories in the hub. The controlled vocabulary is based on the feature list proposed by pseudometa and extended by argentum:

Reviewer 2: ```suggestion
The main goal of adding a list of features is to help compare themes based on what they offer, to save one having to try them out. Keep in mind that themes (especially large ones) have many many features, and it's not our intention to make an exhaustive list of **all** the things your theme can do. Instead, ==ideally the features you list are a way to pitch to (new) users what makes your theme unique==. For that, the idea is to parse the `feature` key to identify some of the frequently sought-after features with a controlled vocabulary and automatically add links to pre-defined categories in the hub. The controlled vocabulary is based on the feature list proposed by pseudometa and extended by argentum:
```

|`highlights` | `[[Themes with custom highlights]]` | Additional colours or ways to achieve the core `==highlight==` behaviour
|`asides` | `[[Themes with side notes]]` | Any features related to the `<aside>` tag or any of its css workarounds, sometimes also called side notes or margin notes
|`checklists` | `[[Themes with custom checklists]]` |Particularly if you added custom types
| `callouts` | `[[Themes with custom callouts]]` | Does the theme add special callouts?
| `css_classes` | `[[Themes with custom css classes]]` | Does the theme add css classes that are not covered on other features? 
| `miscellaneous` | Only for certain keywords (see below)| Any other feature you'd like to mention as a markdown formatted string. Mentioning the following two will add links to the corresponding pages:

Recognised words in the `miscellaneous` features:

- `opinionated` - Will link to `[[Opinionated themes]]`. Whether the theme is opinionated (+ can you opt out of the theme's opinionatedness)
- `minimalistic` -  Will link to [Minimalistic themes](https://publish.obsidian.md/hub/02+-+Community+Expansions/02.02+Themes+by+Category/Minimalistic+Themes) 

As an example:

```yaml
theme:

>     features: 

Reviewer 2: This is being rendered weirdly in Obsidian, I think there may be a mix of tabs and spaces, which is stuffing up the indentation.

Author: Fixed this on the latest commit (I hope... indentation is weird since yaml likes spaces, but the Obsidian outliner plugin prefers tabs).

          custom_callouts:
            docs:
               file: docs/dnd_callouts_overview.md # from root of repo, to be included as is
		       links: # List of strings (can be markdown-formatted)
		           - [More information](www.mytheme.com/dnd_callouts)
            screenshots: # Path to screenshots for this feature (from the root of the repo)
                - file: docs/all_dnd_callouts.png
                  caption: All the D&D callouts for My Theme
                - file: docs/treasure_dnd_callout.png
                  caption: Example of the treasure D&D callout
                  
   ```

%% Note: for the implementation we'd need to know how to structure the templates so that combinations of these make sense %%

### Plugin support

This key (`plugin_support`) is essentially an alternative to parsing the [compatible plugins](https://publish.obsidian.md/hub/04+-+Guides%2C+Workflows%2C+%26+Courses/Guides/Adding+plugin+compatibility+for+themes+to+the+Obsidian+Hub) from the `obsidian.css` file if you prefer to have the hub stuff outside of your theme file. If you don't specify this key, we'll still try to read your plugins from the css file.

```yaml
theme:

>     plugin_support: [] # IDs of the plugins that have been styled by this theme (adding them here would ignore the plugin IDs on the obsidian.css file)

Reviewer 2: Will people know how if strings need to be quoted? I presume they would be comma-separated...

```

### Screenshots

If you want to add additional screenshots, or don't want to add features but prefer to just add screenshots of your theme, you can use the `screenshot` key:

```yaml
theme:
    screenshots: # Path to other screenshots from the root of the repo
        - file: 
          caption:
```

## Template

Here is a full template with all the options being proposed:

```yaml
docs: # Link to the theme documentation
hub_content: [], # List of markdown documents to be included as is to the hub relative to the root of the repo.
authors: # More than one author is possible, start a new line with - 
    - github: # GitHub user names of the people who created the plugin or theme
      websites: [] # Links to personal websites
      discord:
      twitter:
      youtube:
      sponsorships:
          buymeacoffee: # Add username to kofi or buy me a coffee here
          paypal: 
          github:
          patreon:
          others: [] # List of formatted markdown strings

> maintainers: [] # GitHub user names of those maintaining the plugin (leave empty if it's the same as author)

Reviewer: usually, maintainers are also theme devs. not sure whether this is needed?

Author: Yeah, might not be a common practice in themes, but it's something still available (because of plugins). I could remove it from the docs if it makes it confusing.

                - file: 
                  caption:
        publish: true
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:
        gutter:
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:
        print:
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:
        callouts:
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:

>         checklists:

Reviewer: these are more commonly known as custom checkboxes

Author: Should be fixed in 524cc40

                - file: 
                  caption:
        tables: []
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:
        asides: []
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:
        highlights: []
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:
        color_schemes: []
	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:

>         snippets: [] # Path to css snippets relative to the root of the repo. We could parse the css similar to style settings

Reviewer: I think it's worth considering two types of snippets:
1. snippets to enhance/change the theme (requires the user using the theme)
2. snippets to use a certain theme feature in a different theme

Author: Maybe this we can tackle later? Or as a separate categorization of snippets

	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings, e.g. pointing to the css files for download
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:

>         css_classes: []

Reviewer: just for the sake of consistency, I'd write `cssclass`, since the yaml-key in Obsidian also does not use any underscore. (I assume that some people might forget the underscore because they are used to write without it)

Author: Should be fixed in 524cc40

	        docs:
                file: # from root of repo, to be included as is
		        links: [] # markdown-formatted strings
            screenshots: # Path to other screenshots from the root of the repo
                - file: 
                  caption:
        miscellaneous: [] # Other features you want to mention as markdown-formatted strings

>     plugin_support: [] # IDs of the plugins that have been styled by this theme (adding them here would ignore the plugin IDs on the obsidian.css file)

Reviewer: this one I am not too sure what do to about. Problem is that plugin compatibility is only relevant for the 10k club, since they change so much. A simple theme only changing some colors will always be compatible to every plugin, since it's barely different from the default theme.

I think there should be some way to indicate this type of information? 🤔

Author: Maybe it's about phrasing... The plugins listed here should be those that have been tested with/purposefully styled, because yeah, _technically_ all themes are compatible with all plugins.

Reviewer 2: What do we do when the list is present in both CSS and the yaml? I'd prefer it to be only in one place - and if they can list it in the YAML, then that would be the one to retain.
